### PR TITLE
Fields subviews

### PR DESCRIPTION
**Pull request by bartgol**
This PR introduces the possibility to create a field as a "subview" of another. The allowed scenarios are relatively limited, but should be enough to accommodate the use cases we need in scream.

In particular, we can
 -  subview a field by slicing its "leading" dimension (the slowest striding one);
 -  subview a field by slicing its second dimension, IF the input field rank is 3 or more (due to limits in kokkos).
 
 Both these operations can be repeated (so one could strip away, say, the first 3 dimensions). But in all likelihood, only one application of the subview will always be enough in scream.
 
Subviewing is as simple as
```
  // subview f1 at 4th entry along 2nd dim, and use f2_id as field identifier for f2
  auto f2 = f1.subfield(1, 3); // Keeps same name
  auto f3 = f1.subfield("new_name",1,3); // Different field name
```

Important detail (I need to improve the documentation): the view stored in f2 is _the same_ view that is stored in f1. The actual subviewing of it only happens during the call to `get_reshaped_view`. This means that if you do `f2.get_view()` you get the same thing that you would get with `f1.get_view()`. This may be important if `f2` is strided (like in this case, because it subviews f1 along the second dimension). In this case, doing something like
```
Kokkos::deep_copy(f2.get_view()),1.0); // likely not what you expected
```
will fill _the whole original view_ with ones, rather than just filling the subview. In order to get the actual subview, use `get_reshaped_view`:

```
auto v2 = f2.get_reshaped_view<Real***>(); // Will throw if f2 has rank!=3.
Kokkos::deep_copy(v2,1.0); // Ok, only the part of f1 corresponding to f2 gets filled with ones.
```

### CODE REVIEW

**Comment by bartgol** [comment]
This PR is adding the subfield feature, but nowhere in scream (except unit tests) it is used. I would still argue that it is best to merge this now, and work on adding subfield in the FieldRepository (for tracers) in a second PR. I'm afraid I might start to touch quite a bit of code once I get out of the little nutshell where this feature lives.

Also, I need to wait for #E3SM-Project/EKAT81 to be merged, before I can merge this PR (assuming we're ok with updating EKAT within this PR...).

File: components/scream/src/share/tests/dummy_se_point_remapper.hpp
Comment (jeff-cohere) [inline components/scream/src/share/tests/dummy_se_point_remapper.hpp:63]: Is this a meaningful field layout? Or does it indicate an error?

File: components/scream/src/share/tests/dummy_se_point_remapper.hpp
Comment (jeff-cohere) [inline components/scream/src/share/tests/dummy_se_point_remapper.hpp:86]: Same Q as line 63.

File: components/scream/src/share/field/field_tracking.hpp
Comment (jeff-cohere) [inline components/scream/src/share/field/field_tracking.hpp:97]: udpated -> updated

File: components/scream/src/share/field/field_tracking.hpp
Comment (jeff-cohere) [inline components/scream/src/share/field/field_tracking.hpp:58]: worth using `explicit` here?

File: components/scream/src/share/tests/dummy_se_point_remapper.hpp
Comment (bartgol) [reply components/scream/src/share/tests/dummy_se_point_remapper.hpp:63]: The check at the top should throw an error before you hit this case. I think I added it to silence compiler warnings on CUDA maybe? I recall nvcc complaining for a fcn with no return if the default case was hit. I guess some compilers might not be smart enough to figure out that if you'd hit the default case, you'd have already errored out earlier on.

File: components/scream/src/share/field/field_tracking.hpp
Comment (bartgol) [reply components/scream/src/share/field/field_tracking.hpp:58]: Good point.

File: components/scream/src/share/field/field.hpp
Comment (jeff-cohere) [inline components/scream/src/share/field/field.hpp:328]: This is to simplify the implementation in terms of assumptions made to provide padding for layouts, right? Not that it will surprise anyone, but we should probably document this limitation above the declaration of this method.

**Review by jeff-cohere — APPROVED** [review]
There's a lot of stuff here. I think the simplified FieldIdentifiers and the more specific FieldLayouts in the AbstractGrid interface will help people understand things better. I agree we should get these changes in with the knowledge that things will probably evolve a bit as they start being used.

File: components/scream/src/share/field/field.hpp
Comment (bartgol) [reply components/scream/src/share/field/field.hpp:328]: You mean the idim=0,1 constraint? The reason is that if you subview at idim=2+, then there is no way to keep LayoutRight. In Kokkos, LayoutRight allows to store a single stride for the slowest dim, without hurting performance at all (LR computes offsets like `offset = i7 + dim(6)*(i6+dim(5)* (...)) + stride*i0`, where `stride=dim(1)*...*dim(7)` for contiguous views).

If you slice with idim=2, you can still fit in this offset strategy, but if you slice with idim=2+, you can't, and need LayoutStride. Besides the perf loss, I don't think we'd be able to implement it, since the return type would be different.

I can document better in the Field class, where the function is declared. Better yet, make the fcn private, and use a friend free function in field_utils.hpp, with ample documentation (the file field.hpp is too crowded).

**Comment by bartgol** [comment]
@jeff-cohere I added host view capability to Field. In particular:

 - the field stores both a device and a host (1d) view.
 - both `get_view` and `get_reshaped_view` can accept a new template arg, representing the device or mem/exec space where the view is desired (defaults to device, for bwd compatibility).
 - `sync_to_host` and `sync_to_dev` allow to ensure the two are synced.
 - host mirror view is not created until needed, that is, if one of the sync methods is called, or one of the get view methods (with Host space template arg).
  - some internal typedefs have been changed a bit.
 
I also added some meta-utils at the top of the class. Some are a bit more general than just-for-Field, so at some point they might be factored out.

Tests pass on my laptop, I'll let the AT test on our machines.

**Comment by jeff-cohere** [comment]
Looks like the AT machines might have had some trouble configuring SCREAM with the latest changes?

**Comment by bartgol** [comment]
> Looks like the AT machines might have had some trouble configuring SCREAM with the latest changes?

I just figured out the pb (10 seconds ago). It looks like cmake's generator-expression, while recognizing `C_COMPILER_ID` and `CXX_COMPILER_ID`, does not recognize `Fortran_COMPILER_ID`, so this line (in EKAT's `EkatUtils.cmake`) is buggy:

```
  target_compile_options (${targetName} PRIVATE
    $<$<COMPILE_LANGUAGE:Fortran>:$<$<Fortran_COMPILER_ID:GNU>:-w> $<$<Fortran_COMPILER_ID:Intel>: -w>>)
```
I found online workarounds to emulate this line, but I also found that cmake 14+ does recognize that variable. So now I'm wondering whether to check if cmake 3.14 is available or to do the emulation, which would be something like
```
set (Fortran_COMPILER_ID ${CMAKE_Fortran_COMPILER_ID})
...

  target_compile_options (${targetName} PRIVATE
    $<$<COMPILE_LANGUAGE:Fortran>:$<$<STREQUAL:"${Fortran_COMPILER_ID}","GNU">:-w> $<$<STREQUAL:"${Fortran_COMPILER_ID}","Intel"l>: -w>>)
```

**Comment by jeff-cohere** [comment]
What a tangled web we weave... :-)

**Review by jeff-cohere — APPROVED** [review]
Well done! Let's see how this plays in SCREAM! We'll definitely be trying it with EAGLES too, hopefully later in 2021.

**Pull request merged**
